proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -368,6 +368,1382 @@ lemma invoke_sched_context_corres:
+                     \<and> scRefillMax sc' \<le> length (scRefills sc') \<and> scRefills sc' \<noteq> []"
+               in corres_req)
+   apply (prop_tac "sc_relation sc n sc'")
+    apply (fastforce dest: state_relation_pspace_relation)
+   apply (fastforce dest: sc_ko_at_valid_objs_valid_sc'
+                    simp: valid_sched_context'_def sc_relation_def sc_at_pred_n_def obj_at_def
+                          refills_map_def)
+
+  apply (rule_tac Q="\<lambda>_. pspace_aligned and pspace_distinct and sc_at sc_ptr and active_sc_at sc_ptr
+                         and sc_refills_sc_at (\<lambda>refills. 0 < length refills) sc_ptr"
+              and Q'="\<lambda>_. ?conc"
+               in corres_split')
+     apply (rule corres_guard_imp)
+       apply (rule_tac f="\<lambda>sc. sc\<lparr>sc_refills := tl (sc_refills sc)\<rparr>"
+                   and f'="\<lambda>sc''. scRefillCount_update (\<lambda>_. scRefillCount sc' - Suc 0)
+                                   (scRefillHead_update (\<lambda>_. if scRefillHead sc' = scRefillMax sc' - Suc 0
+                                                                then 0
+                                                                else Suc (scRefillHead sc')) sc'')"
+                    in setSchedContext_no_stack_update_corres)
+          apply (prop_tac "tl (sc_refills sc) = drop (Suc 0) (sc_refills sc)")
+           apply (clarsimp simp: sc_relation_def drop_Suc)
+          apply (clarsimp simp: sc_relation_def refills_map_def)
+          apply (rule nth_equalityI)
+           apply (subst length_map)
+           apply (subst length_wrap_slice; simp)
+          apply (subst nth_map)
+           apply (subst length_wrap_slice; simp)
+          apply (subst wrap_slice_index; simp)
+           apply (fastforce split: if_splits)
+          apply (subst wrap_slice_index; simp)
+          apply (clarsimp split: if_splits)
+         apply simp
+        apply (clarsimp simp: objBits_simps)
+       apply simp
+      apply (clarsimp simp: obj_at_def)
+     apply simp
+    apply (fastforce dest: refills_heads_equal)
+   apply wpsimp
+    apply (wpsimp wp: set_object_wp)
+   apply (clarsimp simp: sc_at_pred_n_def obj_at_def is_sc_obj_def vs_all_heap_simps)
+   apply (metis Nitpick.size_list_simp(2) length_ineq_not_Nil(1) less_not_refl3)
+  apply wpsimp
+  apply (fastforce dest: sc_ko_at_valid_objs_valid_sc'
+                   simp: valid_sched_context'_def valid_sched_context_size'_def scBits_simps
+                         objBits_simps
+                  split: if_splits)
+  done
+
+lemma scBits_inverse_sc:
+  assumes "sc_relation sc n sc'"
+  shows "refillAbsoluteMax' (scBitsFromRefillLength' (length (scRefills sc'))) = length (scRefills sc')"
+  using assms
+  apply (simp add: scRefills_length max_num_refills_eq_refillAbsoluteMax')
+  apply (rule arg_cong[where f=refillAbsoluteMax'])
+  apply (rule scBits_inverse_us)
+  by (simp add: sc_const_eq)
+
+lemma refillNew_corres:
+  "1 < max_refills
```

Comment:
I am leaving it as is for now. Since this `max_refills` is an argument for both the abstract and the concrete sides, it probably makes sense to leave it as is, and, without an actual use case, it is not clear at this point which is a better way to go if anything.